### PR TITLE
Add more capabilities to cleanup RunInLooperThread tests

### PR DESCRIPTION
**Pull request by cmelchior**
Breaking out this functionality from #4655 

Also, unit test that tests the unit tests 😜 

@realm/java

### CODE REVIEW

File: realm/realm-library/src/androidTest/java/io/realm/RunTestInLooperThreadLifeCycleTest.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+import io.realm.rule.RunInLooperThread;
+import io.realm.rule.RunTestInLooperThread;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+
+/**
+ * Meta test. Checking the lifecycle of `@RunTestInLooperThreadTest does the right thing.
```
Comment (kneth) [inline realm/realm-library/src/androidTest/java/io/realm/RunTestInLooperThreadLifeCycleTest.java]: Backquote?

File: realm/realm-library/src/androidTest/java/io/realm/rule/RunInLooperThread.java [hunk 2]
```diff
@@ -81,6 +83,14 @@
     // Access guarded by 'lock'
     private List<Realm> testRealms;
 
+    // List of closable resources that will be automatically closed the the test finishes.
```
Comment (kneth) [inline realm/realm-library/src/androidTest/java/io/realm/rule/RunInLooperThread.java]: "the the" -> "when the"

File: realm/realm-library/src/androidTest/java/io/realm/RunTestInLooperThreadLifeCycleTest.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+    @After
+    public void after() {
+        assertTrue(afterCalled.compareAndSet(false, true));
+        assertTrue(testExited.get());
+        assertFalse(looperThread.isTestComplete()); // Beware of this. Use `runAfterTest` for destroying resources used.
+    }
+
+    @Test
+    @RunTestInLooperThread(before = PrepareLooperTest.class)
```
Comment (nhachicha) [inline realm/realm-library/src/androidTest/java/io/realm/RunTestInLooperThreadLifeCycleTest.java]: do we need to assert different states using the `after` runnable as well?

File: realm/realm-library/src/androidTest/java/io/realm/rule/RunInLooperThread.java [hunk 4]
```diff
@@ -290,6 +329,24 @@ private void closeRealms() {
...
+        synchronized (lock) {
+            for (Closeable cr : closableResources) {
+                cr.close();
+            }
+        }
+    }
+
+    /**
+     * Checks if the current test is considered completed or not.
+     * It is completed if either {@link #testComplete()} was called or an uncaught exception was throw.
```
Comment (nhachicha) [inline realm/realm-library/src/androidTest/java/io/realm/rule/RunInLooperThread.java]: was thrown

File: realm/realm-library/src/androidTest/java/io/realm/RunTestInLooperThreadLifeCycleTest.java (hunk 3 above)
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RunTestInLooperThreadLifeCycleTest.java]: Yup. That is being done here https://github.com/realm/realm-java/pull/4740/files#diff-d50b08d92be5fd6f67af340d3b015a63R78

**Pull request merged**